### ISSUE TITLE
parallelUploads does not work with NodeJS

### ISSUE THREAD

**Issue by ViRb3** [tus/tus-js-client#303]
I am using `tux-js-client v2.3.0` with `Node.js v16.13.0`. When uploading with `parallelUploads = 1` and `uploadSize`, as per [the official example](https://github.com/tus/tus-js-client/blob/master/demos/nodejs/index.js), everything works fine. 

However, if you try to use `parallelUploads = 2` or above, the first error you get is:

> tus: cannot use the `uploadSize` option when parallelUploads is enabled

If you remove the option, then this error shows up:

> node:internal/errors:464 01:07
    ErrorCaptureStackTrace(err); 01:07
    ^ 01:07
RangeError [ERR_OUT_OF_RANGE]: The value of "start" is out of range. It must be an integer. Received NaN 01:07
    at new ReadStream (node:internal/fs/streams:193:5) 01:07
    at createReadStream (node:fs:2869:10) 01:07
    at FileSource.slice (REDACTED/node_modules/tus-js-client/lib.es5/node/fileReader.js:168:45) 01:07
    at REDACTED/node_modules/tus-js-client/lib.es5/upload.js:284:31 01:07
    at Array.map (<anonymous>) 01:07
    at Upload._startParallelUpload (REDACTED/node_modules/tus-js-client/lib.es5/upload.js:282:27) 01:07
    at REDACTED/node_modules/tus-js-client/lib.es5/upload.js:251:18 { 01:07
  code: 'ERR_OUT_OF_RANGE' 01:07
}

After some investigation, turns out that the [FileSource](https://github.com/tus/tus-js-client/blob/master/lib/node/sources/FileSource.js) does not expose a `size` property, while the code tries to read it. Because that's undefined, the code then breaks. I worked around this by adding the following line to the constructor:

```diff
class FileSource {
  constructor(stream) {
    this._stream = stream;
    this._path = stream.path.toString();
+   this.size = fs.statSync(this._path).size;
  }
```

Everything works fine after that, however, uploading continues past 100%, and goes up until `parallelUploads * 100%`. I believe that each parallel chunk is actually the full length of the file, not just a part of it as expected.

Any help is much appreciated. 

Sample code:

```js
var fs = require("fs");
var tus = require("tus-js-client");
const { exit } = require("process");

class FileSource {
  constructor(stream) {
    this._stream = stream;
    this._path = stream.path.toString();
    this.size = fs.statSync(this._path).size;
  }
  slice(start, end) {
    const stream = fs.createReadStream(this._path, {
      start,
      end: end - 1,
      autoClose: true,
    });
    stream.size = end - start;
    console.log(start, end, stream.size)
    return Promise.resolve({ value: stream });
  }
  close() {
    this._stream.destroy();
  }
}

class FileReader {
  openFile(input, chunkSize) {
    return Promise.resolve(new FileSource(input));
  }
}

var args = process.argv.slice(2);
if (args.length != 3) {
  console.log("Usage: index.js FILE_TO_UPLOAD ENDPOINT KEY");
  exit(1);
}

var path = args[0];
var endpoint = args[1];
key = args[2];

var file = fs.createReadStream(path);

var options = {
  endpoint: endpoint,
  metadata: {
    filename: "foo",
  },
  headers: {
    Authorization: "Bearer " + key,
  },
  parallelUploads: 4,
  fileReader: new FileReader(),
  onError(error) {
    throw error;
  },
  onProgress(bytesUploaded, bytesTotal) {
    var percentage = ((bytesUploaded / bytesTotal) * 100).toFixed(2);
    console.log(bytesUploaded, bytesTotal, `${percentage}%`);
  },
  onSuccess() {
    console.log("Upload finished!");
  },
};

var upload = new tus.Upload(file, options);
upload.start();

```

**Comment by Acconut** [comment]
> However, if you try to use `parallelUploads = 2` or above, the first error you get is:
> 
> > tus: cannot use the `uploadSize` option when parallelUploads is enabled

Yes, this is working as intended. The `parallelUploads` option is currently not supported in Node.js.

> I worked around this by adding the following line to the constructor:
> 
> ```diff
> +   this.size = fs.statSync(this._path).size;
> ```

You are correct that in order to support `parallelUploads`, the size property is needed. However, in practice we cannot add synchronous file system calls to tus-js-client as that might block the event loop. 

> Everything works fine after that, however, uploading continues past 100%, and goes up until `parallelUploads * 100%`. I believe that each parallel chunk is actually the full length of the file, not just a part of it as expected.

Interesting, I don't know out of my head why the would happen. FileSource has never been tested with parallelUploads before, so it is not unexpected that there is some unwanted behavior.

**Comment by ViRb3** [comment]
Never mind me, I just took another try and after a lot of debugging, created a working `FileSource` implementation. It still doesn't solve the async issue, but it works perfectly.

```diff
class FileSource {
  constructor(stream) {
    this._stream = stream;
    this._path = stream.path.toString();
+   this.size = stream.size ? stream.size : fs.statSync(this._path).size;
+   this.offset = stream.start ? stream.start : 0;
  }
  slice(start, end) {
    const stream = fs.createReadStream(this._path, {
-      start: start,
+      start: start + this.offset,
-      end: end - 1,
+      end: end - 1 + this.offset,
    });
    stream.size = end - start;
    return Promise.resolve({ value: stream });
  }
  close() {
    this._stream.destroy();
  }
}

class FileReader {
  openFile(input, chunkSize) {
    return Promise.resolve(new FileSource(input));
  }
}
```

**Comment by Acconut** [comment]
Nice catch! Congrats on finding the issue! :tada: If you are interested in making this async and contributing it to tus-js-client, feel free to open a PR :)

**Comment by m-cat** [comment]
> Yes, this is working as intended. The parallelUploads option is currently not supported in Node.js.

Should there be an error/warning when one tries to use this option in Node? I just spent a couple hours bumping my head against this -- the error I get is:
```
RangeError [ERR_OUT_OF_RANGE]: The value of "end" is out of range. It must be >= 0 && <= 9007199254740991. Received -1
```

**Comment by Acconut** [comment]
@m-cat Using the patch from @ViRb3 above we maybe can add support for `parallelUploads` in conjunction with streams from `fs.createReadStream`. I haven't tested this enough but it could work. The error you are seeing is not intended. Basically, we have an assertion that `parallelUploads` is not compatible with the `uploadSize` option. We also have an assertion that `fs.createReadStream` requires `uploadSize`, but this is not properly called due to a bug. This is the error message you were looking for.

In short, we should change following things:
- Ensure that we always have an upload size if the `parallelUploads` option is set.
- Automatically obtain upload size if file is passed in using `fs.createReadStream`
- Consider file start when slicing files from ` fs.createReadStream`

I am not sure when I will be coming to these TODOs, but hopefully not too far in the future.

**Comment by ViRb3** [comment]
Happy to say I've been using this patch in production for almost a full year and haven't had a single problem :)

**Comment by Acconut** [comment]
@m-cat @ViRb3 I just implemented all necessary changes and `parallelUploads` is now working with `fs.ReadStream` for me. These improvements will be in the v3 release, which should come out this week.
